<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 15px; border-radius: 10px; text-align: center; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
  Project Title: <span style="color: #333333;">Heart disease analysis</span>
</h1>


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 20px; border-radius: 15px; text-align: center; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
  <span style="color: #1c1c1c; font-size: 2rem;">👨‍💻</span> 
  Author: <span style="color: #333333;">Maaz Umar</span><br>
  <a href="https://github.com/Maaz-Umar-00" style="color: #333333; text-decoration: none; font-size: 0.9rem;">GitHub</a> |
  <a href="https://www.linkedin.com/in/maaz-umar-178090307/" style="color: #333333; text-decoration: none; font-size: 0.9rem;">LinkedIn</a> |
  <a href="mailto:maazumar2005@email.com" style="color: #555555; text-decoration: none; font-size: 0.9rem;">Email</a>

</h1>


<img src="./Heart_pic.webp" alt="Mitron Bank" style="width: 600px; display: block; margin: 0 auto;" />


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 15px; border-radius: 10px; text-align: center; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
  Meta-Data(About Dataset) <span style="color: #333333;"></span>
</h1>



This multivariate dataset includes 16 attributes such as age, sex, chest pain type, blood pressure, cholesterol, and others. Despite having 76 attributes, studies typically use this subset. The Cleveland database is commonly used for predicting heart disease based on these attributes and for gaining insights to better understand the condition.

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
  Columns description: <span style="color: #333333;"></span>
</h1>

* `id:` A unique identifier assigned to each patient. It helps to distinguish between different patients but doesn't provide any medical information.
* `age:` The patient's age in years. Age is a key factor in assessing heart disease risk, as older individuals generally have a higher risk.
* `dataset:` The place or source of the study or data collection. This might be useful for understanding geographic variations in the data.
* `sex:` The patient's gender (Male or Female). Gender can influence the likelihood of developing heart disease, with different risks associated with each gender.
* `cp (chest pain type):` Indicates the type of chest pain experienced by the patient:
  - 1: Typical angina - pain due to heart disease.
  - 2: Atypical angina - pain not necessarily related to heart disease.
  - 3: Non-anginal pain - pain not related to the heart.
  - 4: Asymptomatic - no chest pain experienced.
* `trestbps:` Resting blood pressure (in mm Hg) measured when the patient is admitted to the hospital. High blood pressure is a known risk factor for heart disease.
* `chol:` Serum cholesterol level (in mg/dl). High cholesterol levels can increase the risk of heart disease.
* `fbs:` Indicates whether the patient's fasting blood sugar level is greater than 120 mg/dl (True or False). High fasting blood sugar can be a sign of diabetes, which is a risk factor for heart disease.
* `restecg:` Resting electrocardiographic results:
  - 1: Normal heart activity.
  - 2: ST-T wave abnormality, which can indicate heart issues.
  - 3: Left ventricular hypertrophy, an enlargement of the heart's left ventricle.
* `thalach:` Maximum heart rate achieved during exercise. A higher heart rate can indicate better cardiovascular fitness, but extremely high or low rates can be a concern.
* `exang:` Exercise-induced angina (True or False). Indicates whether exercise triggers chest pain, which can be a sign of heart disease.
* `oldpeak:` ST depression caused by exercise relative to rest. ST depression can be a sign of heart disease.
* `slope:` The slope of the peak exercise ST segment:
  - 1: Upsloping - indicates better exercise tolerance.
  - 2: Flat - can be normal or indicative of heart issues.
  - 3: Downsloping - often associated with heart disease.
* `ca:` Number of major vessels (0-3) colored by fluoroscopy. More vessels colored indicates more significant coronary artery disease.
* `thal:` Thalassemia status.
* `num:` Indicates whether heart disease is present:
  - 0: No heart disease.
  - 1: 1, 2, and 4 indicate varying levels of disease presence.



<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
  Project Aim: <span style="color: #333333;"></span>
</h1>


The aim of this project is to perform data analysis on the heart disease dataset to understand key factors that may influence the risk of heart disease.The focus will be on the following:

* Understanding the distribution of patient demographics like age and gender.
* Identifying risk factors associated with heart disease, such as high cholesterol, blood pressure, and chest pain types.
* Comparing the presence of heart disease across different subgroups, including age ranges, gender, and other medical attributes.
* Visualizing relationships between variables like cholesterol levels, blood pressure, and heart disease to provide insights into patient health.
* `The ultimate goal is to provide meaningful insights from the dataset that could be useful for understanding heart disease risk.` 


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
Key Findings: <span style="color: #333333;"></span>
</h1>

* Most people in the dataset are between the `ages of 40 and 70.`
* There are` 565 males` in the dataset, which is significant since heart disease is more prevalent in males. The total number of `females is 183`.
* `Middle-aged` and` older` individuals have the` highest rates of heart disease`, which is expected, as the likelihood of developing heart disease increases with age.
* Heart disease is significantly more prevalent in` males, with 317 cases, compared to 40 cases in females.`
* Most people exhibit asymptomatic chest `pain`, yet they are still diagnosed with heart disease. This underscores the importance of thorough medical evaluations and monitoring, even for patients with non-specific or no symptoms at all.
* It was found that as `age increases, both resting blood pressure and cholesterol levels tend to rise.` This could be one of the reasons why `middle-aged and older individuals have a higher risk of heart disease.`
* As age increases, maximum heart rate (thalch) values tend to decrease. This suggests potential concerns related to` cardiovascular fitness and health`, indicating that older individuals may experience reduced heart efficiency and a higher risk of heart-related issues.
* `The final and most important finding was a strong correlation between' high blood pressure, high cholesterol, elevated heart rate, and severe depression in the oldpeak variable. This suggests that individuals with these characteristics are significantly more likely to have heart disease. The data emphasizes the importance of monitoring these health indicators to effectively assess heart disease risk.


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 15px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
  1. Importing Libraries <span style="color: #333333;"></span>
</h1>

Let's start the project by importing all the libraries that we will need in this project.


In [1]:
# import libraries

# 1. to handle the data
import pandas as pd
import numpy as np

# to visualize the dataset
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# To preprocess the data and scales the data
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder

# for imputation of missing values
from sklearn.impute import SimpleImputer, KNNImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

# For little bit of  machine learning
from sklearn.linear_model import LinearRegression

# ignore warnings   
import warnings
warnings.filterwarnings('ignore')

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
  1.1 Loading the dataset using pandas library <span style="color: #333333;"></span>
</h1>

In [2]:
# loading the data from csv file placed locally in our pc
df = pd.read_csv('heart_disease_uci.csv')

# printing the first 5 rows of the dataframe
df.head()

id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num  
0       fixed defect    0  
1             normal    2  
2  reversable defect    1  
3             normal    0  
4             normal    0

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
 1.2 Checking total number of columns and rows <span style="color: #333333;"></span>
</h1>

In [3]:
print("Total number of rows are ",df.shape[0])
print("Total number of columns  are ",df.shape[1])

Total number of rows are  920
Total number of columns  are  16


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
1.3 Checking basic info and statistics of the data <span style="color: #333333;"></span>
</h1>

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        920 non-null    int64  
 1   age       920 non-null    int64  
 2   sex       920 non-null    object 
 3   dataset   920 non-null    object 
 4   cp        920 non-null    object 
 5   trestbps  861 non-null    float64
 6   chol      890 non-null    float64
 7   fbs       830 non-null    object 
 8   restecg   918 non-null    object 
 9   thalch    865 non-null    float64
 10  exang     865 non-null    object 
 11  oldpeak   858 non-null    float64
 12  slope     611 non-null    object 
 13  ca        309 non-null    float64
 14  thal      434 non-null    object 
 15  num       920 non-null    int64  
dtypes: float64(5), int64(3), object(8)
memory usage: 115.1+ KB


In [5]:
df.describe()

id         age    trestbps        chol      thalch     oldpeak  \
count  920.000000  920.000000  861.000000  890.000000  865.000000  858.000000   
mean   460.500000   53.510870  132.132404  199.130337  137.545665    0.878788   
std    265.725422    9.424685   19.066070  110.780810   25.926276    1.091226   
min      1.000000   28.000000    0.000000    0.000000   60.000000   -2.600000   
25%    230.750000   47.000000  120.000000  175.000000  120.000000    0.000000   
50%    460.500000   54.000000  130.000000  223.000000  140.000000    0.500000   
75%    690.250000   60.000000  140.000000  268.000000  157.000000    1.500000   
max    920.000000   77.000000  200.000000  603.000000  202.000000    6.200000   

               ca         num  
count  309.000000  920.000000  
mean     0.676375    0.995652  
std      0.935653    1.142693  
min      0.000000    0.000000  
25%      0.000000    0.000000  
50%      0.000000    1.000000  
75%      1.000000    2.000000  
max      3.000000    4.000000

# **Observations**
---

1. There are **920 rows** and **16 columns** in the dataset.
2. The columns in the dataset are:
   - `'id', 'age', 'sex', 'dataset', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 
      'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'`
3. Some columns have missing values, and we will explore these missing values in detail later in this notebook.
   - For example, `trestbps` has **861 non-null** values, `chol` has **890 non-null** values, `ca` has only **309 non-null** values, etc.
4. After using the `describe()` function, the following summary statistics were observed:
   - The minimum value for `age` is **28**, and the maximum is **77**.
   - The average resting blood pressure (`trestbps`) is **132.13 mm Hg**, with a minimum of **0** and a maximum of **200 mm Hg**.
   - The average cholesterol (`chol`) level is **199.13 mg/dl**, with a minimum of **0** and a maximum of **603 mg/dl**.
   - The average maximum heart rate (`thalch`) is **137.55 bpm**, with a minimum of **60 bpm** and a maximum of **202 bpm**.
   - The `oldpeak` values (ST depression induced by exercise) range from **-2.6** to **6.2**.
 


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 15px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
2. Before proceeding with further analysis on the `dataset` we will first clean the dataset. Data cleaning is essential to ensure that we are working with accurate and complete data. <span style="color: #333333;"></span>
</h1>



<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
 2.1 First, check if the data types are correct before proceeding further.. <span style="color: #333333;"></span>
</h1>


In [6]:
df.dtypes

id            int64
age           int64
sex          object
dataset      object
cp           object
trestbps    float64
chol        float64
fbs          object
restecg      object
thalch      float64
exang        object
oldpeak     float64
slope        object
ca          float64
thal         object
num           int64
dtype: object

- All data types are correct; let's move on to the next step. If any issues arise, we'll address them then

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
2.2 Now,Checking for missing values <span style="color: #333333;"></span>
</h1>


### 


In [7]:
missing_percentage = round((df.isnull().sum() / len(df) * 100), 2).sort_values(ascending=False)
print(f'Percentage of missing values')
print(missing_percentage)

Percentage of missing values
ca          66.41
thal        52.83
slope       33.59
fbs          9.78
oldpeak      6.74
trestbps     6.41
thalch       5.98
exang        5.98
chol         3.26
restecg      0.22
id           0.00
age          0.00
sex          0.00
dataset      0.00
cp           0.00
num          0.00
dtype: float64


- Visualizing the null values


In [8]:
import plotly.express as px

# Calculating the percentage of missing values for each column
missing_percentage = df.isnull().mean() * 100

# Converting to DataFrame for plotting
missing_percentage_df = missing_percentage.reset_index()
missing_percentage_df.columns = ['Column', 'Missing Percentage']

# Creating the bar plot
fig = px.bar(missing_percentage_df, x='Column', y='Missing Percentage', title='Missing Values Percentage by Column')

# Updating layout with height, title formatting, axis titles, and width
fig.update_layout(
    height=600,  
    width=1200,  
    title={
        'text': 'Missing Values Percentage by Column',
        'font': {'size': 24, 'family': 'Arial'}, 
        'x': 0.5,  
        'xanchor': 'center'
    },
    xaxis_title='Columns',
    yaxis_title='Percentage (%)'
)

# Adding labels to the bars
fig.update_traces(
    text=missing_percentage_df['Missing Percentage'], 
    texttemplate='%{text:.2f}%', 
    textposition='outside'
)

# Showing the plot
fig.show()


- The high percentage of missing values in the `ca 66.41`and `thal 52.41` columns is significant enough to potentially impact the quality and reliability of our analysis. Therefore, we will drop these columns to ensure that our analysis is based on more reliable data.
- For the remaining columns with lower missing value percentages, we will proceed with imputation to handle any missing values:

- `Dropping ca and thal column`

In [9]:
# Droping columns with high missing values
df.drop(columns=['ca', 'thal'], inplace=True)


- `Imputing missing values using mean for numerical column and mode for categorical columns and for slope column i will use an Machine Learning model.`

In [10]:
# List of numerical columns to impute with mean
numerical_columns = ['oldpeak', 'trestbps', 'thalch', 'chol']

# List of categorical columns to impute with mode
categorical_columns = ['fbs', 'exang','restecg']

# Imputing numerical columns with mean
num_imputer = SimpleImputer(strategy='mean')
df[numerical_columns] = num_imputer.fit_transform(df[numerical_columns])

# Imputing categorical columns with mode
cat_imputer = SimpleImputer(strategy='most_frequent')
df[categorical_columns] = cat_imputer.fit_transform(df[categorical_columns])

* Lets Used an `K-Neirest-Neighbour Machine learning Algorithim` to impute missing values in `slope column`.

In [11]:


# Step 1: Manually mapping categorical values to numerical values
slope_mapping = {'flat': 0, 'upsloping': 1, 'downsloping': 2}
df['slope_mapped'] = df['slope'].map(slope_mapping)

# Step 2: Replacing NaN with np.nan for the KNN Imputer
df['slope_mapped'] = df['slope_mapped'].replace(np.nan, np.nan)

# Step 3: Initializing and apply KNN Imputer
knn_imputer = KNNImputer(n_neighbors=3)
df[['slope_mapped']] = knn_imputer.fit_transform(df[['slope_mapped']])

# Step 4: Maping the numerical values back to the original categories
reverse_mapping = {0: 'flat', 1: 'upsloping', 2: 'downsloping'}
df['slope'] = df['slope_mapped'].round().map(reverse_mapping)

# Step 5: Droping the mapped column because it's no longer needed
df.drop(columns=['slope_mapped'], inplace=True)

# Verifing the imputation worked
print(df['slope'].value_counts(dropna=False))


slope
upsloping      512
flat           345
downsloping     63
Name: count, dtype: int64


* `Now Verifying the data`

In [12]:
df.isnull().sum()

id          0
age         0
sex         0
dataset     0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalch      0
exang       0
oldpeak     0
slope       0
num         0
dtype: int64

### Summary of Actions Taken for Missing Values
---
1. `Dropped Columns:` We removed the ca and thal columns due to their high percentage of missing values.

2. `Imputed Missing Values:`\
**`Numerical Columns: `Replaced missing values with the` mean.`**\
**`Categorical Columns:` Replaced missing values with the `mode.`**

3. `Machine Learning Imputation for slope:` Used the `K-Nearest Neighbors (KNN)imputer` to fill missing values, leveraging its ability to predict missing entries based on similar data points.

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
2.3 Checking for outliers and incorrect values <span style="color: #333333;"></span>
</h1>



In [89]:
# Dictionary to map columns to descriptive titles and axis names
descriptive_titles = {
    'chol': {'title': 'Cholesterol Level', 'yaxis': 'Cholesterol (mg/dL)', 'xaxis': 'Index'},
    'trestbps': {'title': 'Resting Blood Pressure', 'yaxis': 'Blood Pressure (mm Hg)', 'xaxis': 'Index'},
    'age': {'title': 'Age of Patient', 'yaxis': 'Age (Years)', 'xaxis': 'Index'},
    'thalch': {'title': 'Maximum Heart Rate', 'yaxis': 'Heart Rate (bpm)', 'xaxis': 'Index'},
    'oldpeak': {'title': 'Depression Induced by Exercise', 'yaxis': 'Oldpeak (Depression Level)', 'xaxis': 'Index'}
}

# List of specific columns to visualize
columns_to_plot = ['chol', 'trestbps', 'age', 'thalch', 'oldpeak']

# Visualizing outliers using box plots for each specified column
for column in columns_to_plot:
    fig = px.box(df, y=column, title=descriptive_titles[column]['title'])
    fig.update_layout(
        yaxis_title=descriptive_titles[column]['yaxis'],
        xaxis_title=descriptive_titles[column]['xaxis'],
        title={'text': descriptive_titles[column]['title'], 'x': 0.5, 'xanchor': 'center'}
    )
    fig.show()


* Dropping rows where cholestrol level and resting blood pressure level is 0

In [13]:
# Droping rows where cholesterol level is 0
df = df[df['chol'] != 0]

# Droping rows where resting blood pressure is 0
df = df[df['trestbps'] != 0]

# Printing the shape of the cleaned data
print("Data after dropping rows with zero values:", df.shape)


Data after dropping rows with zero values: (748, 14)


## Observations:
----
After analyzing the plots, we found that in these two columns:

##### Cholesterol Level (chol):
* Min: 0 mg/dL
* Max: 603 mg/dL
##### Resting Blood Pressure (trestbps):
* Min: 0 mm Hg
* Max: 200 mm Hg
So We are dropped zero values because it is `biologically impossible.`

* Cholesterol: 0 mg/dL is `biologically impossible.`
* Blood Pressure: 0 mm Hg is also `medically impossible.`

`All other columns show no anomalies or outliers, ensuring data integrity for the analysis.`








<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 10px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
2.4 Feature Engineering <span style="color: #333333;"></span>
</h1>


In [14]:
df.head()

id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope  num  
0  lv hypertrophy   150.0  False      2.3  downsloping    0  
1  lv hypertrophy   108.0   True      1.5         flat    2  
2  lv hypertrophy   129.0   True      2.6         flat    1  
3          normal   187.0  False      3.5  downsloping    0  
4  lv hypertrophy   172.0  False      1.4    upsloping    0

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 8px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1rem;">
2.4.1 Creating New Features Based on Ranges of `Resting Blood Pressure (trestbps)` to Enhance Our Analysis. <span style="color: #333333;"></span>
</h1>

* Normal: 90–120 mm Hg
* Elevated: 120–130 mm Hg
* Hypertension Stage 1: 130–140 mm Hg
* Hypertension Stage 2: >140 mm Hg

In [15]:
def categorize_blood_pressure(value):
    if value < 90:
        return 'Low'
    elif 90 <= value <= 120:
        return 'Normal'
    elif 120 < value <= 130:
        return 'Elevated'
    elif 130 < value <= 140:
        return 'Hypertension Stage 1'
    else:
        return 'Hypertension Stage 2'

df['blood_pressure_category'] = df['trestbps'].apply(categorize_blood_pressure)


* Now doing the same things for `cholestorl level(chol)`,`Heart rate Zones(thalch)` and `oldpeak`.

In [16]:
def categorize_cholesterol(value):
    if value < 200:
        return 'Normal'
    elif 200 <= value < 240:
        return 'Borderline High'
    else:
        return 'High'

df['cholesterol_category'] = df['chol'].apply(categorize_cholesterol)

def categorize_heart_rate(value):
    if value < 60:
        return 'Low'
    elif 60 <= value <= 100:
        return 'Moderate'
    else:
        return 'High'

df['heart_rate_category'] = df['thalch'].apply(categorize_heart_rate)

def categorize_oldpeak(value):
    if value == 0:
        return 'No Depression'
    elif 0 < value <= 1:
        return 'Mild Depression'
    elif 1 < value <= 2:
        return 'Moderate Depression'
    else:
        return 'Severe Depression'

df['oldpeak_category'] = df['oldpeak'].apply(categorize_oldpeak)



In [17]:
df.head()

id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope  num  \
0  lv hypertrophy   150.0  False      2.3  downsloping    0   
1  lv hypertrophy   108.0   True      1.5         flat    2   
2  lv hypertrophy   129.0   True      2.6         flat    1   
3          normal   187.0  False      3.5  downsloping    0   
4  lv hypertrophy   172.0  False      1.4    upsloping    0   

  blood_pressure_category cholesterol_category heart_rate_category  \
0    Hypertension Stage 2      Borderline High                High   
1    Hypertension Stage 2                 High                High   
2                  Normal      Borderline High                High   
3                Elevated                 High                High   
4                Elevated      Borderline High                High   

      oldpeak_category  
0    Severe Depression  
1  Moderate Depression  
2    Severe Depression  
3    Severe Depression  
4  Moderate Depression

## Summary of Steps in Feature Engineering:

#### 1. Blood Pressure Categories (`trestbps`)
- **Normal**: 90–120 mm Hg
- **Elevated**: 120–130 mm Hg
- **Hypertension Stage 1**: 130–140 mm Hg
- **Hypertension Stage 2**: >140 mm Hg
#### 2. Cholesterol Levels (`chol`)
- **Normal**: < 200 mg/dL
- **Borderline High**: 200–239 mg/dL
- **High**: ≥ 240 mg/dL
#### 3. Heart Rate Zones (`thalch`)
- **Low**: < 60 bpm
- **Moderate**: 60–100 bpm
- **High**: > 100 bpm
#### 4. Oldpeak Interpretation (`oldpeak`)
- **No Depression**: 0
- **Mild**: 0–1
- **Moderate**: 1–2
- **Severe**: > 2
##### Summary of New Features
- `blood_pressure_category`
- `cholesterol_category`
- `heart_rate_category`
- `oldpeak_category`

These features enhance the analysis of heart disease risk.


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 8px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1rem;">
2.4.2 I Have checked  the unique values in the num column, which include 0, 1, 2, and 4.<span style="color: #333333;"></span>
</h1>

* 0 means "No disease."
* 1, 2, and 4 indicate varying levels of disease presence.
* I will assign values based on these conditions: if a value is 0, I will write "No disease." If a value is greater than 0, I will write "Yes, disease present."

In [18]:
df['num'] = df['num'].apply(lambda x: 'No disease' if x == 0 else 'Yes disease')

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: #f7f7f7; padding: 20px; border-radius: 12px; text-align: center; box-shadow: 0 4px 12px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 2rem; margin-top: 30px;">
The End! Thank you for Reading.
</h1><h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: #f7f7f7; padding: 15px; border-radius: 12px; text-align: center; box-shadow: 0 4px 12px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem; margin-top: 30px;">
I Hope you found this project insightful and informative.
</h1>